docker/generate_makefile.py

```diff
@@ -0,0 +1,327 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Simple generator for local Makefile rules."""
+
+import os
+import yaml
+
+BASE_TAG = 'gcr.io/fuzzbench'
+BENCHMARKS_DIR = os.path.join(os.path.dirname(__file__), os.pardir,
+                              'benchmarks')
+FUZZERS_DIR = os.path.join(os.path.dirname(__file__), os.pardir, 'fuzzers')
+
+BOILERPLATE = """
+cache_from = $(if ${RUNNING_ON_CI},--cache-from {fuzzer},)
+"""
+
+FUZZER_TEMPLATE = """
+.{fuzzer}-builder: base-builder
+	docker build \\
+    --tag {base_tag}/builders/{fuzzer} \\
+    --file fuzzers/{fuzzer}/builder.Dockerfile \\
+    $(call cache_from,{base_tag}/builders/{fuzzer}) \\
+    fuzzers/{fuzzer}
+
+.pull-{fuzzer}-builder: pull-base-builder
+	docker pull {base_tag}/builders/{fuzzer}
+"""
+
+FUZZER_BENCHMARK_TEMPLATE = """
+.{fuzzer}-{benchmark}-builder: .{fuzzer}-builder
+	docker build \\
+    --tag {base_tag}/builders/{fuzzer}/{benchmark} \\
+    --build-arg fuzzer={fuzzer} \\
+    --build-arg benchmark={benchmark} \\
+    $(call cache_from,{base_tag}/builders/{fuzzer}/{benchmark}) \\
+    --file docker/benchmark-builder/Dockerfile \\
+    {build_arguments} \\
```

Comment:
I dont understand how this will just work, are these arguments to fuzzer.py or to docker build ?

Reply:
+1

Would it be possible to do https://github.com/google/fuzzbench/pull/342#discussion_r426094426, i.e.:

have rules like:

{fuzzer}-{variant}-{benchmark}

and just pass:

--build-arg fuzzer=$(1) \
--build-arg variant=$(2) \
--build-arg benchmark=$(3) \

Reply:
Sorry for the late reply here, Abhishek and I discussed this in a VC last week. It shouldn't be necessary to separate out the variant from the fuzzer name. The variant names must already be unique based on how other pieces of this are implemented. In other parts of the code we handle things in a similar way.

As for the first question (just to have it documented) this will work because when a variant is defined we have both a fuzzer_environment section for environment variables set before calling into fuzzer.py (already implemented and not relevant to this change) and a separate build_arguments section, which is just a set of docker arguments passed to the builder. This was the most generic way we could handle things, and doesn't limit us to environment variables only. If we decide that we want to use the environment in both cases it would be a fairly small change and I'm open to it, but Jonathan suggested handling it this way a while back.

Reply:
Thanks!

(Left another comment on the script where `build_argument` is created to help readers understand how this works)